Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,484 @@
+#include <LibGUI/TableView.h>
+#include <LibGUI/TreeView.h>
+#include <LibIMAP/QuotedPrintable.h>
+
+MailWidget::MailWidget()
+{
+    load_from_gml(mail_window_gml);
+
+    m_mailbox_list = *find_descendant_of_type_named<GUI::TreeView>("mailbox_list");
+    m_individual_mailbox_view = *find_descendant_of_type_named<GUI::TableView>("individual_mailbox_view");
+    m_web_view = *find_descendant_of_type_named<Web::OutOfProcessWebView>("web_view");
+    m_statusbar = *find_descendant_of_type_named<GUI::Statusbar>("statusbar");
+
+    m_mailbox_list->on_selection_change = [this] {
+        selected_mailbox();
+    };
+
+    m_web_view->on_link_click = [this](auto& url, auto&, unsigned) {
+        if (!Desktop::Launcher::open(url)) {
+            GUI::MessageBox::show(
+                window(),
+                String::formatted("The link to '{}' could not be opened.", url),
+                "Failed to open link",
+                GUI::MessageBox::Type::Error);
+        }
+    };
+
+    m_web_view->on_link_middle_click = [this](auto& url, auto& target, unsigned modifiers) {
+        m_web_view->on_link_click(url, target, modifiers);
+    };
+
+    m_web_view->on_link_hover = [this](auto& url) {
+        if (url.is_valid())
+            m_statusbar->set_text(url.to_string());
+        else
+            m_statusbar->set_text("");
+    };
+
+    m_link_context_menu = GUI::Menu::construct();
+    auto link_default_action = GUI::Action::create("&Open in Browser", [this](auto&) {
+        m_web_view->on_link_click(m_link_context_menu_url, "", 0);
+    });
+    m_link_context_menu->add_action(link_default_action);
+    m_link_context_menu_default_action = link_default_action;
+    m_link_context_menu->add_separator();
+    m_link_context_menu->add_action(GUI::Action::create("&Copy URL", [this](auto&) {
+        GUI::Clipboard::the().set_plain_text(m_link_context_menu_url.to_string());
+    }));
+
+    m_web_view->on_link_context_menu_request = [this](auto& url, auto& screen_position) {
+        m_link_context_menu_url = url;
+        m_link_context_menu->popup(screen_position, m_link_context_menu_default_action);
+    };
+
+    m_image_context_menu = GUI::Menu::construct();
+    m_image_context_menu->add_action(GUI::Action::create("&Copy Image", [this](auto&) {
+        if (m_image_context_menu_bitmap.is_valid())
+            GUI::Clipboard::the().set_bitmap(*m_image_context_menu_bitmap.bitmap());
+    }));
+    m_image_context_menu->add_action(GUI::Action::create("Copy Image &URL", [this](auto&) {
+        GUI::Clipboard::the().set_plain_text(m_image_context_menu_url.to_string());
+    }));
+    m_image_context_menu->add_separator();
+    m_image_context_menu->add_action(GUI::Action::create("&Open Image in Browser", [this](auto&) {
+        m_web_view->on_link_click(m_image_context_menu_url, "", 0);
+    }));
+
+    m_web_view->on_image_context_menu_request = [this](auto& image_url, auto& screen_position, const Gfx::ShareableBitmap& shareable_bitmap) {
```

Comment:
east-const please :)

```suggestion
    m_web_view->on_image_context_menu_request = [this](auto& image_url, auto& screen_position, Gfx::ShareableBitmap const& shareable_bitmap) {
```